### PR TITLE
refactor: swap

### PR DESCRIPTION
**Pull request by nuanyang233**
image

### CODE REVIEW

**Comment by wenluomask** [comment]
1. Max feature doesn't work, it is unresponsible;

2. Switch to other network tabs, the balance or swap option belongs to current network;
截屏2021-11-11 下午2 45 46

3. Missing estimate tx cost in $;

4. Missing new gas fee setting page and unlock info;
截屏2021-11-11 下午2 57 21

5. Don't show all the swap exchange options initially, only show the best one. And once the best option appears, select the best option automatically;

6. Stop loading if user didn't choose any token to swap;
截屏2021-11-11 下午3 02 24

7. It usually takes a long time to choose the best choice;

8. For the insufficient liquidity swap pair, mask didn't show the message but give an incorrect estimate amount;
截屏2021-11-11 下午3 02 24
截屏2021-11-11 下午3 09 52

9. Didn't provide Exact Unlock option to user, user will infinite unlock the token;
截屏2021-11-11 下午3 16 28

10. Before you unlock the token, the swap button should be disable to click;

11. Make the gas fee and buttons below always fix position;

12. It is supposed to show insufficient balance or no gas fee messgae if balance is not enough to finish current swap;

13. UI adjustment:
- All parts need to align to stright line;
- The token icon size doesn't match the Figma design;
截屏2021-11-11 下午3 52 58
- Use same blue color for everywhere needed, MingMing said it is a variable that fetch from Twitter;
截屏2021-11-11 下午3 28 52

**Comment by wenluomask** [comment]
15.Estimate tx is inconsistent with app and actual cost;
截屏2021-11-17 下午3 48 00
截屏2021-11-17 下午3 44 34

16. Select Token button's color is not follow Twitter system;
截屏2021-11-17 下午3 36 49

17. Click the arrow to switch tokens' order;
截屏2021-11-17 下午3 46 52

18. Unlock info's UI is not match with Figma;
截屏2021-11-17 下午3 56 48

**Comment by wenluomask** [comment]
25. The unlock tip should be appeared by hover over, now will always appear;

26. Crash happened when open confirmation page;(#4880 )
Screen Shot 2021-11-22 at 4 52 15 PM

**Comment by wenluomask** [comment]
- [ ] 27. Switch network button issues:
- [ ] - If reject to switch network from MetaMask, the button should back to switch network button(blue color), but not green switch success button;
Screen Shot 2021-11-23 at 1 42 52 PM

- [ ] -For switching status, only keep three spot ..., currently there are six;
Screen Shot 2021-11-23 at 1 43 21 PM

- [ ] 28. For the confirm swap page:
- [ ] -From and To token amount should be align to middle; 
Screen Shot 2021-11-23 at 2 18 39 PM
- [ ] -Delete the frame of scroll bar and change to dark color in dark mode;
- [ ] -Price should be the format 1 Token = xx Token; 
Screen Shot 2021-11-23 at 2 19 09 PM

- [ ] 29. Let Unlock info appear above the icon;
Screen Shot 2021-11-23 at 2 46 01 PM

- [ ] 30. The tx cost change to Gas fee, and the amount is incorrect currently;
Screen Shot 2021-11-23 at 2 50 38 PM
Screen Shot 2021-11-23 at 2 50 26 PM

- [ ] 31. I have changed gas price before, but it didn't show the change gas in confirm page;
Screen Shot 2021-11-23 at 3 10 14 PM
Screen Shot 2021-11-23 at 3 10 26 PM

- [ ] 32. Change close icon to back icon for swap page;

Screen Shot 2021-11-23 at 3 20 17 PM

- [ ] 33. Swap in 0x need to confirm extra places and execution reverted errors occurred;
Screen Shot 2021-11-23 at 2 49 38 PM

- [ ] -34. If swap successfully, there should be a message pop out;

- [ ] -35. Didn't choose the best choice in Arbitrum and xDai network;
Screen Shot 2021-11-23 at 3 32 51 PM

**Comment by wenluomask** [comment]
Close https://mask.atlassian.net/browse/MASK-289?jql=labels%20%3D%20%222.2.0%22

Checked.

**Comment by wenluomask** [comment]
Checked.

File: packages/mask/src/extension/popups/pages/Wallet/GasSetting/GasSetting1559.tsx [hunk 1]
```diff
@@ -237,11 +236,14 @@ export const GasSetting1559 = memo(() => {
...
-                    formatWeiToGwei(new BigNumber(value.computedPayload._tx.maxPriorityFeePerGas, 16)).toString(),
+                    fromWei(
+                        new BigNumber(value.computedPayload._tx.maxPriorityFeePerGas).toString(),
+                        'gwei',
+                    ).toString(),
                 )
                 setValue(
                     'maxFeePerGas',
-                    formatWeiToGwei(new BigNumber(value.computedPayload._tx.maxFeePerGas, 16)).toString(),
+                    fromWei(new BigNumber(value.computedPayload._tx.maxFeePerGas).toString(), 'gwei').toString(),
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/GasSetting/GasSetting1559.tsx:246]:
```suggestion
                    fromWei(new BigNumber(value.computedPayload._tx.maxFeePerGas).toFixed(), 'gwei').toString(),
```

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/ConfirmDialog.tsx [hunk 2]
```diff
@@ -67,42 +114,161 @@ export function ConfirmDialogUI(props: ConfirmDialogUIProps) {
     }, [trade, executionPrice])
     //#endregion
 
+    //#region gas price
+    const nativeToken = createNativeToken(chainId)
+    const tokenPrice = useNativeTokenPrice(chainId)
+
+    const gasFee = useMemo(() => {
+        return gas && gasPrice ? new BigNumber(gasPrice).multipliedBy(gas).integerValue().toFixed() : 0
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trader/ConfirmDialog.tsx:122]: To return the same value type.

```suggestion
        return gas && gasPrice ? new BigNumber(gasPrice).multipliedBy(gas).integerValue().toFixed() : '0'
```

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/ConfirmDialog.tsx [hunk 3]
```diff
@@ -67,42 +114,161 @@ export function ConfirmDialogUI(props: ConfirmDialogUIProps) {
...
+    }, [gas, gasPrice])
+
+    const feeValueUSD = useMemo(
+        () => (gasFee ? new BigNumber(formatWeiToEther(gasFee).times(tokenPrice).toFixed(2).toString()) : '0'),
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trader/ConfirmDialog.tsx:126]: Duplicate `toString` and `toFixed`.

```suggestion
        () => (gasFee ? new BigNumber(formatWeiToEther(gasFee).times(tokenPrice).toFixed(2)) : '0'),
```

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/Gas1559Settings.tsx [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+
+    //#region Confirm function
+    const handleConfirm = useCallback(
+        (data: zod.infer<typeof schema>) => {
+            onSave_({
+                maxPriorityFeePerGas: toHex(formatGweiToWei(data.maxPriorityFeePerGas).toString()),
+                maxFeePerGas: toHex(formatGweiToWei(data.maxFeePerGas).toString()),
+            })
+        },
+        [onSave_],
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trader/Gas1559Settings.tsx:231]:
```suggestion
        [onSave_, data.maxPriorityFeePerGas, data. maxFeePerGas],
```

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/Gas1559Settings.tsx [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+        return undefined
+    }, [maxFeePerGas, gasOptions, getGasOptionsLoading])
+    //endregion
+
+    //#region Confirm function
+    const handleConfirm = useCallback(
+        (data: zod.infer<typeof schema>) => {
+            onSave_({
+                maxPriorityFeePerGas: toHex(formatGweiToWei(data.maxPriorityFeePerGas).toString()),
+                maxFeePerGas: toHex(formatGweiToWei(data.maxFeePerGas).toString()),
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trader/Gas1559Settings.tsx:228]:
```suggestion
                maxPriorityFeePerGas: toHex(formatGweiToWei(data.maxPriorityFeePerGas).toFixed()),
                maxFeePerGas: toHex(formatGweiToWei(data.maxFeePerGas).toFixed()),
```

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/Gas1559Settings.tsx [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+
+    const onSave = handleSubmit(handleConfirm)
+    //#endregion
+
+    //#region If the selected changed, set the value on the option to the form data
+    useEffect(() => {
+        if (selected !== null) {
+            setValue(
+                'maxPriorityFeePerGas',
+                new BigNumber(options[selected].content?.suggestedMaxPriorityFeePerGas ?? 0).toString() ?? '',
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trader/Gas1559Settings.tsx:242]:
```suggestion
                new BigNumber(options[selected].content?.suggestedMaxPriorityFeePerGas ?? 0).toFixed() ?? '',
```

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/Gas1559Settings.tsx [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+            )
+            setValue(
+                'maxFeePerGas',
+                new BigNumber(options[selected].content?.suggestedMaxFeePerGas ?? 0).toString() ?? '',
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trader/Gas1559Settings.tsx:246]:
```suggestion
                new BigNumber(options[selected].content?.suggestedMaxFeePerGas ?? 0).toFixed() ?? '',
```

File: packages/mask/src/plugins/Trader/trader/uniswap/usePools.ts [hunk 8]
```diff
@@ -45,17 +49,25 @@ export function usePools(
...
-    const poolContracts = usePoolContracts(poolAddresses)
+    const poolContracts = usePoolContracts(poolAddresses, chainId)
+
+    const { value: targetBlockNumber } = useTargetBlockNumber(chainId)
 
     const [slot0s, slot0sCalls, slot0sState, slot0sCallback] = useMultipleContractSingleData(
         poolContracts,
         Array.from<'slot0'>({ length: poolContracts.length }).fill('slot0'),
         [],
+        1_000_000,
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/trader/uniswap/usePools.ts:60]: Avoid dead code magic numbers. Alter the signature to:

```ts
function useMultipleContractSingleData<T extends BaseContract, K extends keyof T['methods']>(
    contracts: T[],
    names: K[],
    callData: Parameters<T['methods'][K]>,
    chainId = currentChainid.value,
    blockNumber = currentBlockNumber.value,
    gasLimit = DEFAULT_GAS_LIMIT,
)
```

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/Gas1559Settings.tsx (hunk 4 above)
Comment (nuanyang233) [reply packages/mask/src/plugins/Trader/SNSAdaptor/trader/Gas1559Settings.tsx:231]: The `data` is param of function

File: packages/web3-shared/evm/hooks/useMulticall.ts [hunk 9]
```diff
@@ -142,6 +144,7 @@ export function useMulticallCallback() {
                     results: calls.map((call) => getCallResult(call, chainId, blockNumber) ?? [false, '0x0', '0x0']),
                 })
             } catch (error) {
+                console.log(error)
```
Comment (guanbinrui) [inline packages/web3-shared/evm/hooks/useMulticall.ts:147]:
```suggestion
```

File: packages/mask/src/plugins/Wallet/services/tokenPrice.ts [hunk 10]
```diff
@@ -18,6 +18,7 @@ function updateCurrentPrices(data: CryptoPrice) {
             }
         })
     })
+    console.log('update', currentPrices)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Wallet/services/tokenPrice.ts:21]:
```suggestion
```

File: packages/mask/src/plugins/Wallet/services/chain.ts [hunk 11]
```diff
@@ -38,6 +51,13 @@ export async function updateChainState() {
                     ? getBalance(currentAccountSettings.value, {
                           chainId: currentChainIdSettings.value,
                           providerType: currentProviderSettings.value,
+                      }).then((value) => {
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Wallet/services/chain.ts:54]: Does the `getBalance` below also need this logic?

Comment (nuanyang233) [reply packages/mask/src/plugins/Wallet/services/chain.ts:54]: Here is to update the cache value after each balance is updated

File: packages/mask/src/components/shared/ApplicationBoard.tsx [hunk 12]
```diff
@@ -37,7 +37,7 @@ const useStyles = makeStyles()((theme) => ({
         minHeight: 36,
         margin: '0 auto',
         borderRadius: 4,
-        backgroundColor: theme.palette.background.default,
+        // backgroundColor: theme.palette.background.default,
```
Comment (guanbinrui) [inline packages/mask/src/components/shared/ApplicationBoard.tsx:40]:
```suggestion
```

File: packages/mask/src/components/shared/ApplicationBoard.tsx [hunk 13]
```diff
@@ -131,7 +131,8 @@ export function ApplicationBoard({ secondEntries, secondEntryChainTabs }: MaskAp
     //#endregion
 
     //#region Swap
-    const { openDialog: openSwapDialog } = useRemoteControlledDialog(PluginTraderMessages.swapDialogUpdated)
+    // const { openDialog: openSwapDialog } = useRemoteControlledDialog(PluginTraderMessages.swapDialogUpdated)
```
Comment (guanbinrui) [inline packages/mask/src/components/shared/ApplicationBoard.tsx:134]:
```suggestion
```

**Pull request merged**